### PR TITLE
[1822Africa] Implementation p2

### PR DESCRIPTION
**Pull request by vandamm**
- [x] Branch is derived from the latest `master`
- [x] Code passes linter with `docker compose exec rack rubocop -a`
- [x] Tests pass cleanly with `docker compose exec rack rake`

### Implementation Notes

* **Explanation of Change**

Continuation of #9414. Progress can be tracked in #9419

### CODE REVIEW

File: lib/engine/game/g_1822_africa/game.rb [hunk 1]
```diff
@@ -252,35 +335,47 @@ def setup_companies
...
+          head = @companies[0...next_concession_index]
+          tail = @companies[next_concession_index..-1]
 
-        def put_concession_to_front(companies)
-          first_concession_index = companies.find_index { |c| c.id[0] == self.class::COMPANY_CONCESSION_PREFIX }
+          @companies = tail + head
 
-          head = companies[0...first_concession_index]
-          tail = companies[first_concession_index..-1]
+          unowned_head = head.filter { |c| !c.owner || c.owner == @bank }
```
Comment (michaeljb) [inline lib/engine/game/g_1822_africa/game.rb:359]:
```suggestion
          unowned_head = head.select { |c| !c.owner || c.owner == @bank }
```

File: lib/engine/game/g_1822_africa/game.rb [hunk 2]
```diff
@@ -292,57 +387,128 @@ def timeline
...
+        def stock_round(round_num = 1)
+          G1822Africa::Round::Stock.new(self, [
+            Engine::Step::DiscardTrain,
+            G1822::Step::BuySellParShares,
+          ], round_num: round_num)
+        end
+
+        def total_rounds(name)
+          @operating_rounds if name == self.class::OPERATING_ROUND_NAME
+          2 if name == self.class::STOCK_ROUND_NAME
```
Comment (michaeljb) [inline lib/engine/game/g_1822_africa/game.rb:451]: In ORs this method will return `nil`.

This should use a `case name` block. https://rubystyle.guide/#case-vs-if-else

File: lib/engine/game/g_1822_africa/game.rb [hunk 3]
```diff
@@ -292,57 +387,128 @@ def timeline
...
-          # End if bid boxes cannot be refilled AFTER SR
+        def compute_game_end
+          return %i[custom full_or] if bidbox.length < self.class::BIDDING_BOX_MINOR_COUNT
+          return %i[stock_market current_or] if @stock_market.max_reached?
         end
 
         # Temporary stub
         def setup_exchange_tokens; end
 
         # Stubbed out because this game doesn't it, but base 22 does
```
Comment (michaeljb) [inline lib/engine/game/g_1822_africa/game.rb:508]:
```suggestion
        # Stubbed out because this game doesn't use it, but base 22 does
```

**Review by michaeljb — CHANGES_REQUESTED** [review]
Thanks for the contribution!

As stated in the default PR description, when implementing a new game it's best to keep the changes smaller. I don't mind seeing boiler-plate type things grouped together (so entities, new icons, constant values, base 1822 overrides would have all been fine), but it would have been better for the new logic implementing the special rules of this title to have been split out into other PRs.

No need to restructure that stuff right now, but keep that in mind in the future. I'll merge when my other comments are addressed.

File: lib/engine/game/g_1822_africa/step/dividend.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  module Game
+    module G1822Africa
+      module Step
+        class Dividend < Engine::Game::G1822::Step::Dividend
+          def share_price_change(entity, revenue = 0)
+            return { share_direction: :left, share_times: 1 } unless revenue.positive?
+
+            price = entity.share_price.price
+
+            times = (revenue.to_f / price).floor
```
[the remark this answers is not in the document]
Comment (scottredracecar) [reply lib/engine/game/g_1822_africa/step/dividend.rb:15]: Minors work like other 1822 games, always pay 50/50, always move up (or down) exactly one space

**Comment by vandamm** [comment]
@michaeljb thanks for the review! I’m sorry for submitting such a big PR, but I’ve tried to keep it structured per commit. I’ll make sure to stick to smaller ones in the future. 
I’ll fix your remarks tomorrow.

File: lib/engine/game/g_1822_africa/game.rb (hunk 2 above)
Comment (vandamm) [reply lib/engine/game/g_1822_africa/game.rb:451]: Sorry for this silly mistake! I learned some  Ruby specifically to contribute here so I may miss some conventions or make mistakes :)

Comment (michaeljb) [reply lib/engine/game/g_1822_africa/game.rb:451]: If you do ever have a function similar to this but the `if`s are checking different conditions, what you have would be fine if the first line just started with a `return`. You could make both/all lines start with a `return` and let rubocop clean up the last `return`.

File: lib/engine/game/g_1822_africa/game.rb (hunk 1 above)
Comment (vandamm) [reply lib/engine/game/g_1822_africa/game.rb:359]: I'm changing it though I don't quite get the difference, they look identical to me

**Review by michaeljb — APPROVED** [review]
Looks good, thanks!

**Pull request merged**